### ISSUE TITLE
systemd: our ExecReload command doesn't wait for the reload to complete

### ISSUE THREAD

**Issue by Al2Klimov** [Icinga/icinga2#7301]
[systemd.service(5)](https://www.freedesktop.org/software/systemd/man/systemd.service.html#ExecReload=) clearly says:

"It is strongly recommended to set ExecReload= to a command that not only triggers a configuration reload of the daemon, but also synchronously waits for it to complete."

**Comment by dnsmichi** [comment]
The lines before are more interesting:

> One additional, special environment variable is set: if known, $MAINPID is set to the main process of the daemon, and may be used for command lines like the following:
>
> /bin/kill -HUP $MAINPID
> Note however that reloading a daemon by sending a signal (as with the example line above) is usually not a good choice, because this is an asynchronous operation and hence not suitable to order reloads of multiple services against each other.

Which basically means that systemd invented their own method for reloads, sort of deprecating the HUP signal always being asynchronous.

Thing is, we cannot just drop the HUP signal - there's still systems where Icinga runs with SysVInit, and likewise, inside containers where signals are fired "just like this".

I'd be interested how the systemd developers and users have solved this with other prominent applications.

**Comment by Al2Klimov** [comment]
I've patched our script like this:

```
while kill -0 $pid >/dev/null 2>&1; do
        sleep 1
done
```

... but this isn't a complete solution.

**Comment by dnsmichi** [comment]
@Elias481 might this be related to the SIGBUS crashes you're seeing during reloads?

**Comment by Elias481** [comment]
@dnsmichi I don't think this one is exactly related to #7132. From what I understand is the reason for this recommendation only to make the unit "suitable to order reloads of multiple services against each other" so to be able to reload multiple services one after the other without expensive/complex waiting for the previous service to finish reloading. I cannot see any connection to the reload procedure within application.
But I don't know the library for sd_notify, it could be possible that the sd_notify somehow invoved in the SIGBUS and that that in turn is related to that. (But I think it should't, if it would it would be some bug in systemd.)

**Comment by Al2Klimov** [comment]
@dnsmichi By the way: Why are we validating our config inside this script?

**Comment by carraroj** [comment]
ref/NC/690026
